### PR TITLE
Improvements to contextual functions

### LINKED ISSUE #4764: Complete set of cur_ functions
And cross-reference with data.table in docs

*  `cur_data()`       = `.SD`
* `group_keys()`     -> `cur_group()`      = `.BY`
* `group_indices()`  -> `cur_group_id()`   = `.GRP`
* `group_rows()`     -> `cur_group_loc()`  = `.I`

### PR DESCRIPTION
**Pull request by hadley**
* New `cur_data()` and `cur_group_rows()
* Connect contextual functions to data.table pronouns
* Access `cur_group_id()` and `n()` through mask instead of context

Fixes #4764

### CODE REVIEW

File: R/conditions.R [hunk 1]
```diff
@@ -142,28 +139,26 @@ stop_mutate_not_vector <- function(result, index, dots) {
 
   name <- arg_name(dots, index)
   expr <- as_label(quo_get_expr(dots[[index]]))
-  data <- peek_mask()$full_data()
-  group <- peek_mask()$get_current_group()
+  data <- context_get("..mask")$full_data()
```
Comment (romainfrancois) [inline R/conditions.R]: Should this perhaps be `cur_mask()` even if not exported ?

File: src/dplyr.h [hunk 2]
```diff
@@ -76,8 +74,6 @@ SEXP rows_i = VECTOR_ELT(rows, i);
 R_xlen_t n_i = XLENGTH(rows_i);                                                      \
 SEXP current_group = PROTECT(Rf_ScalarInteger(i + 1));                               \
 Rf_defineVar(dplyr::symbols::current_group, current_group, env_private);             \
```
Comment (romainfrancois) [inline src/dplyr.h]: Then I think we can directly use: 

```c
Rf_defineVar(dplyr::symbols::current_group, Rf_ScalarInteger(i + 1), env_private);
```

File: R/conditions.R (hunk 1 above)
Comment (hadley) [reply R/conditions.R]: Yeah, I still feel like this needs a bit more refactoring. I'll have a go at making it cleaner.

File: src/dplyr.h (hunk 2 above)
Comment (hadley) [reply src/dplyr.h]: Will do.

File: R/conditions.R (hunk 1 above)
Comment (hadley) [reply R/conditions.R]: What if we got rid of the private variables in DataMask? That would reduce the method surface of the class, and then we could have `cur_mask("data")`, `cur_mask("keys")` etc

Comment (hadley) [reply R/conditions.R]: Hmmm, that's probably too much for this PR. I'll just bring back `cur_mask()`

Comment (hadley) [reply R/conditions.R]: Ok, bought back `peek_mask()` and did some refactoring so the mask and current column can share more code

**Pull request merged**